librz/core/rop.c

```diff
@@ -414,41 +441,79 @@ RZ_API bool rz_core_rop_gadget_info_has_register(const RZ_NONNULL RzRopGadgetInf
 	return false;
 }
 
-static bool is_var_write_event(const RzRopRegInfo *reg_info) {
+static inline bool is_var_read_event(const RzRopRegInfo *reg_info) {
+	return reg_info->is_var_read && !reg_info->is_mem_read && !reg_info->is_var_write && !reg_info->is_mem_write;
+}
+
+static inline bool is_var_write_event(const RzRopRegInfo *reg_info) {
 	return reg_info->is_var_write;
 }
 
-static bool is_mem_read_event(const RzRopRegInfo *reg_info) {
+static inline bool is_mem_read_event(const RzRopRegInfo *reg_info) {
 	return reg_info->is_mem_read;
 }
 
-static bool is_mem_write_event(const RzRopRegInfo *reg_info) {
+static inline bool is_mem_write_event(const RzRopRegInfo *reg_info) {
 	return reg_info->is_mem_write;
 }
 
-static bool is_pc_write_event(const RzRopRegInfo *reg_info) {
+static inline bool is_pc_write_event(const RzRopRegInfo *reg_info) {
 	return reg_info->is_pc_write;
 }
 
 rz_rop_event_check_fn rz_rop_event_functions[RZ_ROP_EVENT_COUNT] = {
+	is_var_read_event,
 	is_var_write_event,
 	is_mem_read_event,
 	is_mem_write_event,
 	is_pc_write_event,
 };
 
+/**
+ * \brief Check if a given event dependency is present for a register
+ * \param gadget_info RZ_NONNULL Pointer to the RzRopGadgetInfo object.
+ * \param event The RzRopEvent to check.
+ * \param reg_name Name of the register
+ * \return True if there is an event which \p event which uses \p reg_name. False otherwise.
+ */
+RZ_API bool rz_core_rop_gadget_reg_info_has_event(const RZ_NONNULL RzRopGadgetInfo *gadget_info,
+	const RzRopEvent event, const RZ_NULLABLE char *reg_name) {
+	rz_return_val_if_fail(gadget_info, NULL);
+	if (event < 0 || event >= RZ_ROP_EVENT_COUNT) {
+		return false;
+	}
+	RzListIter *iter;
+	RzRopRegInfo *reg_info;
+	rz_list_foreach (gadget_info->dependencies, iter, reg_info) {
+		if (RZ_STR_NE(reg_name, reg_info->name)) {
+			continue;
+		}
+		if (event == RZ_ROP_EVENT_COUNT) {
+			return true;
+		}
+		if (rz_rop_event_functions[event](reg_info)) {
+			return true;
+		}
+	}
+	return false;
+}
+
 /**
  * \brief Find all dependencies based on a specific event in the dependencies of a RzRopGadgetInfo object.
  * \param gadget_info RZ_NONNULL Pointer to the RzRopGadgetInfo object.
  * \param event The RzRopEvent to check.
  * \return RZ_OUT A pointer to a list of RzRopRegInfo objects matching the given event, or NULL if none are found or if gadget_info is NULL.
  */
-RZ_API RzPVector /*<RzRopRegInfo *>*/ *rz_core_rop_gadget_get_reg_info_by_event(const RZ_NONNULL RzRopGadgetInfo *gadget_info, const RzRopEvent event) {
+RZ_API RZ_NULLABLE RzPVector /*<RzRopRegInfo *>*/ *rz_core_rop_gadget_get_reg_info_by_event(const RZ_NONNULL RzRopGadgetInfo *gadget_info,
```

Comment:
Ownership of the retval.

Reply:
https://github.com/rizinorg/rizin/pull/4582#discussion_r1730545228